contracts/tokenfunds.js

```diff
@@ -0,0 +1,601 @@
+      if (api.assert(approval !== null, 'you have not approved this proposal')) {
+        await api.db.remove('approvals', approval);
+
+        const balance = await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: dtf.voteToken });
+        let approvalWeight = 0;
+        if (balance && balance.stake) {
+          approvalWeight = balance.stake;
+        }
+        if (balance && balance.delegationsIn) {
+          approvalWeight = api.BigNumber(approvalWeight)
+            .plus(balance.delegationsIn)
+            .toFixed(voteTokenObj.precision, api.BigNumber.ROUND_HALF_UP);
+        }
+        const wIndex = acct.weights.findIndex(x => x.symbol === dtf.voteToken);
+        if (wIndex !== -1) {
+          acct.weights[wIndex].weight = approvalWeight;
+        } else {
+          acct.weights.push({ symbol: dtf.voteToken, weight: approvalWeight });
+        }
+        await api.db.update('accounts', acct);
+        await updateProposalWeight(proposal._id, api.BigNumber(approvalWeight).negated());
+        api.emit('disapproveProposal', { id: proposal._id });
+      }
+    }
+  }
+};
+
+actions.updateProposalApprovals = async (payload) => {
+  const { account, token, callingContractInfo } = payload;
+
+  if (callingContractInfo === undefined) return;
+  if (callingContractInfo.name !== 'tokens') return;
+
+  const acct = await api.db.findOne('accounts', { account });
+  if (acct !== null) {
+    const params = await api.db.findOne('params', {});
+
+    // calculate approval weight of the account
+    const balance = await api.db.findOneInTable('tokens', 'balances', { account, symbol: token.symbol });
+    let approvalWeight = 0;
+    if (balance && balance.stake) {
+      approvalWeight = balance.stake;
+    }
+
+    if (balance && balance.delegationsIn) {
+      approvalWeight = api.BigNumber(approvalWeight)
+        .plus(balance.delegationsIn)
+        .toFixed(token.precision, api.BigNumber.ROUND_HALF_UP);
+    }
+
+    const wIndex = acct.weights.findIndex(x => x.symbol === token.symbol);
+    let oldApprovalWeight = 0;
+    if (wIndex !== -1) {
+      oldApprovalWeight = acct.weights[wIndex].weight;
+      acct.weights[wIndex].weight = approvalWeight;
+    } else {
+      acct.weights.push({ symbol: token.symbol, weight: approvalWeight });
+    }
+
+    const deltaApprovalWeight = api.BigNumber(approvalWeight)
+      .minus(oldApprovalWeight)
+      .dp(token.precision, api.BigNumber.ROUND_HALF_UP);
+
+    if (!api.BigNumber(deltaApprovalWeight).eq(0)) {
+      await api.db.update('accounts', acct);
+      const approvals = await api.db.find('approvals',
+        { from: account },
+        params.maxAccountApprovals,
+        0,
+        [{ index: '_id', descending: true }]);
+      for (let index = 0; index < approvals.length; index += 1) {
+        const approval = approvals[index];
+        await updateProposalWeight(approval.to, deltaApprovalWeight);
```

Comment:
I'm still a bit concerned about performance in this loop.  maxAccountApprovals is 1000, so in the worst case we might call updateProposalWeight 1000 times, and that function has a both a DB find & update, so again in the worst case that's 2000 more queries, though most likely the validation check against the proposal's endDate will cut down the updates significantly.

You could optimize it a bit by copying the endDate into the approvals table when the approval is created.  That way the information will already be available here so you can do the endDate validation without performing another DB find if the proposal has already ended.  Although if you made that change, you'd probably also want to remove the ability to update the endDate through the updateProposal action, since that would then require updating it in all the approvals table entries as well, which could have its own performance implications.  So there are trade-offs.

Also, maybe a smaller number for maxAccountApprovals would be appropriate?  1000 seems like a lot, perhaps 100 or even 50 would be better?

Reply:
In my experience with the hived/DHF there have been quite a few requests for reducing the end date via an update. Proposal creators gain more flexibility if they are unsure whether they will get funded or not, and after successful funding can adjust amount and end to meet their goals without having to create a new one. It would be good to try and keep it.

An advantage of going with fewer though, such as 30 or 50, should encourage people to approve only what they believe to be higher quality proposals. It is designed to limit approvals on active proposals at any given time, so instead of having updateProposal do a large query to do the update, I am trying to update them as they go along in approveProposal and updateProposalApprovals.

Also noticed here we need to enforce the max when people are adding new approvals, to prevent any stake against an active proposal to not be updated. So I need to add more queries here, but they offer a chance to update approvals on finished proposals and verify user is not going over the max amount. If people come against this error when approving, then it will be time to consider increasing maxAccountApprovals.

So, going with 50, in either case I think it would be a max of 50 + 50 queries. (The endDate validation now doesn't save a query,  but causes the approval to be updated instead, but allows the whole thing to remain consistent).